subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1895,3 +1836,134 @@ int ztls_setsockopt(int sock, int level, int optname,
 
 	return 0;
 }
+
+static ssize_t tls_sock_read_vmeth(void *obj, void *buffer, size_t count)
+{
+	return ztls_recvfrom_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static ssize_t tls_sock_write_vmeth(void *obj, const void *buffer,
+				    size_t count)
+{
+	return ztls_sendto_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static int tls_sock_ioctl_vmeth(void *obj, unsigned int request, ...)
+{
+	switch (request) {
+	case ZFD_IOCTL_CLOSE:
+		return ztls_close_ctx(obj);
+
+	case ZFD_IOCTL_FCNTL: {
+		va_list args;
+		int err;
+
+		/* Pass the call to the core socket implementation. */
+		va_start(args, request);
+		err = sock_fd_op_vtable.fd_vtable.ioctl(obj, request, args);
```

Comment:
@rlubos: Sorry, I didn't notice this before. What is happening here? did you test this? Let me conjecture this isn't correct. You pass a `va_list` here, but what va_list is an effectively a pointer to first argument. But `...` in ioctl() means an argument *value*, not address.

So, while there's almost certainly an issue, the way to fix it may be actually to make ->ioctl() take `va_list` as the last argument, not `...`. Because you code shows a common problem with variadic functions: how to dispatch them, preserving the original args? The only portable solution would be to convert `...` to `va_list` and pass `va_list` around.

Reply:
@pfalcon Right, giving it a second look I see it's not correct. I"ve ran the only sample that use `fcntl` (echo_async) with TLS socket and it worked so I assumed solutions works, but yeah apparently that's not correct.

I can try to fix it in a way you described, yet only converting `...` to `va_list` doesn't seem to be enough. We still want a code like this to work:
https://github.com/zephyrproject-rtos/zephyr/blob/c012a288c371c69420ea25c3bb7bb9af5a743f07/lib/fdtable.c#L202
Shall we provide a generic wrapper to ioctl which would dispatch to vtable and call it there? Something like this:
```
int ioctl(int fildes, int request, ... /* arg */)
{
    ...
    va_start(args, request);

    return fdtable[fd].vtable->ioctl(fdtable[fd].obj, request, args);
}
```

Reply:
@rlubos: I didn't mention that I'm working on all this stuff, so please don't spend cycles on that, but please be prepared to review.

Reply:
Ack, thanks for clarifying this!

Reply:
A hasty WIP PR, just to show progress is posted: https://github.com/zephyrproject-rtos/zephyr/pull/12002